api/envoy/type/metadata/v2/metadata.proto

```diff
@@ -0,0 +1,61 @@
+syntax = "proto3";
+
+package envoy.type.metadata.v2;
+
+option java_outer_classname = "MetadataProto";
+option java_multiple_files = true;
+option java_package = "io.envoyproxy.envoy.type.metadata.v2";
+
+import "validate/validate.proto";
+
+// [#protodoc-title: Metadata]
+
+// MetadataKey provides a general interface using `filter` and `path` to retrieve value from
+// :ref:`Metadata <envoy_api_msg_core.Metadata>`.
+//
+// For example, for the following Metadata:
+//
+// .. code-block:: yaml
+//
+//    filter_metadata:
+//      envoy.filters.http.rbac:
+//        path:
+//          to:
+//            foo: bar
+//            people:
+//              name: bob
+//
+// The following MetadataKey is matched as the path ``[path, to, foo]``
+// will retrieve a string value "bar" from the Metadata.
+//
+// .. code-block:: yaml
+//
+//    filter: envoy.filters.http.rbac
+//    path:
+//    - key: path
+//    - key: to
+//    - key: foo
+//
+message MetadataKey {
+  // Specifies the segment in a path to retrieve value from Metadata.
+  // Currently it is only supported to specify the key, i.e. field name, as one segment of a path.
+  message PathSegment {
+    oneof segment {
+      option (validate.required) = true;
+
+      // If specified, use the key to retrieve the value in a Struct.
+      string key = 1 [(validate.rules).string = {min_bytes: 1}];
+    }
+  }
+
+  // The filter name to retrieve the Struct from the Metadata.
```

Comment:
IIRC we have multiple places that store metadata, right? Routes, hosts, clusters, maybe more? (I actually don't know all the places). Should we have this type be agonstic of where the metadata itself actually comes from and thus remove filter from this? And instead potentially be more specific about which metadata to pull from in the place where it is used? Or do we need to somehow encode into this type what type of metadata we are looking at?

Reply:
Sorry for delay.
Did you mean that, metadata key is only consisted with the full path. Then it is like a general proto struct key other than metadata.
In this tracing scenario, we should define the filter in Metadata type custom tag, i.e.

```
message Metadata {
  string filter = 1;
  ..MetadataKey metadata_key = 2;
  string default_value;
}
```

I'm not professional about it, so not quit sure based on the current metadata structure :), WDYT @htuch @kyessenov @objectiser ?

Reply:
Sorry my point is that AIUI, `Metadata` can be attached to an endpoint, route, cluster, and maybe other things. How does this structure tell us where to actually pull the metadata from for the tag?

Reply:
Maybe we need to rethink the model given how many different kinds of metadata exist, but I can give the original idea and we can see if that makes sense. We basically made the key be a filter or pseudo-filter, e.g. `envoy.lb` for upstream metadata. This destination was essentially a named mail box where metadata is posted _to_. So, with some well known names, this can easily work, e.g. data destined for a WAF filter would be written to `com.acme.waf`, which might be further specialized with additional qualifiers, e.g. `com.acme.waf.route`, as needed. The namespace is global, but there is clearly some amount of specialization happening, e.g. upstream metadata is only available in some contexts.

Reply:
TBH this whole metadata thing is very confusing to me. I get the original intention, but if metadata can come from more than one place, either there has to be a well known way of resolving all of this places into a single thing (for e.g. populating a tag) or the user needs to specify where to pull the metadata from?

Reply:
For tags, I think we might want to be explicit. In many other situations, the metadata matcher has this source contextually implied. E.g. for subset LB, it's clear that the metadata match in the route is being applied to the endpoint metadata.

Reply:
+1. @yittg can you take a look at this a bit more? We probably need a way to specify where the tag population should pull the metadata from?

Reply:
I provided two different field for different metadata sources for now.
And yeah, i will have another full look about metadata in envoy and encode the metadata source to the `CustomTag.Metadata` type.

Reply:
Given that we are now extracting metadata from multiple locations, should this field be called `key_name` or similar? @htuch?

Reply:
I think the idea of metadata is it still has a destination of sorts. This is typically some filter or major subsystem. If we go to key name, we need to make clear the cases that this is really a filter, rather than some made up entity.

Reply:
@htuch sorry I don't follow. Stepping back from a user perspective, this feature is allowing me to add custom spans that pull metadata from e host, route, etc. I don't think there is really any "filter" here per-say. So shouldn't we make this more generic in terms of a "name", "slot_name", etc.?

Reply:
I'm also confused about the "filter" :). Intuitively, it's a namespace, and it's a definition of the metadata, will not only be used by this feature. A general "namespace" or "domain" may be ok.. or "partition" intended to means a subsystem?🤣

Reply:
Yeah, probably we need to start moving away from filter, but I rather we have some idea of how we are going to do naming. I think we need to have something in the developer docs explaining this concept. Basically, to me, we're describing a namespace that:
1. Is acting as a rendezvous; someone is producing data in this namespace, someone is consuming.
2. Needs a consistent naming scheme. I think that for builtins in Envoy, a `envoy.foo` prefix with well-known name makes sense. For custom extensions, I would continue to advise on the use of reverse DNS, e.g. `com.acme.foo`.
3. In the situation where there are multiple producers and multiple consumers, there needs to be some standard for naming and format. An example of this today is auth data that is consumed by RBAC and other other filters.